src/coreComponents/mesh/CellElementSubRegion.hpp

```diff
@@ -247,13 +329,76 @@ class CellElementSubRegion : public CellBlock
     } );
   }
 
+  void calculateElementGeometricQuantities( NodeManager const & nodeManager,
+                                            FaceManager const & faceManager ) override;
+
+private:
+
   /// Map used for constitutive grouping
   map< string, localIndex_array > m_constitutiveGrouping;
 
   /// Array of constitutive point volume fraction
   array3d< real64 > m_constitutivePointVolumeFraction;
 
-private:
+  /// Element-to-node relation
+  NodeMapType m_toNodesRelation;
+
+  /// Element-to-edge relation
+  EdgeMapType m_toEdgesRelation;
+
+  /// Element-to-face relation
+  FaceMapType m_toFacesRelation;
+
+  /// Name of the properties registered from an external mesh
+  string_array m_externalPropertyNames;
+
+  /**
+   * @brief Compute the volume of the k-th element in the subregion.
+   * @param[in] k the index of the element in the subregion
+   * @param[in] X an arrayView of (const) node positions
+   */
+  inline void calculateCellVolumesKernel( localIndex const k,
+                                          arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const & X ) const
+  {
+    LvArray::tensorOps::fill< 3 >( m_elementCenter[ k ], 0 );
+
+    real64 Xlocal[10][3];
+
+    for( localIndex a = 0; a < m_numNodesPerElement; ++a )
+    {
+      LvArray::tensorOps::copy< 3 >( Xlocal[ a ], X[ m_toNodesRelation( k, a ) ] );
+      LvArray::tensorOps::add< 3 >( m_elementCenter[ k ], X[ m_toNodesRelation( k, a ) ] );
+    }
+    LvArray::tensorOps::scale< 3 >( m_elementCenter[ k ], 1.0 / m_numNodesPerElement );
+
+    switch( m_elementType )
+    {
+      case ElementType::Hexahedron:
+      {
+        m_elementVolume[k] = computationalGeometry::HexVolume( Xlocal );
+        break;
+      }
+      case ElementType::Tetrahedron:
+      {
+        m_elementVolume[k] = computationalGeometry::TetVolume( Xlocal );
+        break;
+      }
+      case ElementType::Prism:
+      {
+        m_elementVolume[k] = computationalGeometry::WedgeVolume( Xlocal );
+        break;
+      }
+      case ElementType::Pyramid:
+      {
+        m_elementVolume[k] = computationalGeometry::PyramidVolume( Xlocal );
+        break;
+      }
+      default:
+      {
+        GEOSX_ERROR( "Volume calculation not supported for element type: " << m_elementType );
```

Comment:
Maybe add a more explicit error message specifying the class and the name of the object?

Reply:
done